blink/biosimage.c

```diff
@@ -122,9 +128,41 @@ static const u8 defbios[] = {
                   0xEB, 0xFA),                               // jmp .-4
     BIOS_BYTES_AT(kBiosEnd - 1, 0x00)};
 
-void LoadDefaultBios(struct Machine *m) {
-  size_t kBiosSize = sizeof(defbios);
-  memcpy(m->system->real + kBiosEnd - kBiosSize, defbios, kBiosSize);
+void LoadBios(struct Machine *m, const char *biosprog) {
+  off_t size;
+  if (biosprog) {
+    off_t kBiosMinSize = kBiosEnd - kBiosEntry,
+          kBiosMaxSize = kBiosEnd - kBiosOptBase;
+    int fd;
+    struct stat st;
+    char tmp[64];
+    if ((fd = VfsOpen(AT_FDCWD, biosprog, O_RDONLY, 0)) == -1 ||
+        VfsFstat(fd, &st) == -1) {
+      WriteErrorString(biosprog);
+      WriteErrorString(": failed to load alternate BIOS (errno ");
+      FormatInt64(tmp, errno);
+      WriteErrorString(tmp);
+      WriteErrorString(")\n");
+      exit(127);
+    } else if ((size = st.st_size) < kBiosMinSize) {
+      WriteErrorString(biosprog);
+      WriteErrorString(": failed to load alternate BIOS (file too small)\n");
+    } else if (size > kBiosMaxSize) {
+      WriteErrorString(biosprog);
+      WriteErrorString(": failed to load alternate BIOS (file too large)\n");
+    } else if (VfsRead(fd, m->system->real + kBiosEnd - size, size) != size) {
```

Comment:
The above two error cases should also exit, rather than displaying an error that will likely immediately be overwritten with the TUI.

Reply:
@ghaerr : whoops.  I have fixed this.  Thanks!